packages/controllers/src/cronjob/CronjobController.ts

```diff
@@ -0,0 +1,342 @@
+ * Use this controller to register and schedule periodically executed jobs
+ * using RPC method hooks.
+ */
+export class CronjobController extends BaseController<
+  typeof controllerName,
+  CronjobControllerState,
+  CronjobControllerMessenger
+> {
+  private _messenger: CronjobControllerMessenger;
+
+  private _timers: Map<string, Timer>;
+
+  // Mapping from jobId to snapId
+  private _snapIds: Map<string, string>;
+
+  private _dailyTimer!: Timer;
+
+  constructor({ messenger, state }: CronjobControllerArgs) {
+    super({
+      messenger,
+      metadata: {
+        jobs: { persist: true, anonymous: false },
+      },
+      name: controllerName,
+      state: {
+        jobs: {},
+        ...state,
+      },
+    });
+    this._timers = new Map();
+    this._snapIds = new Map();
+    this._messenger = messenger;
+    this.dailyCheckin();
+
+    this._handleEventSnapInstalled = this._handleEventSnapInstalled.bind(this);
+    this._handleEventSnapRemoved = this._handleEventSnapRemoved.bind(this);
+    this._handleEventSnapUpdated = this._handleEventSnapUpdated.bind(this);
+
+    // Subscribe to Snap events
+    this.messagingSystem.subscribe(
+      'SnapController:snapInstalled',
+      this._handleEventSnapInstalled,
+    );
+
+    this.messagingSystem.subscribe(
+      'SnapController:snapRemoved',
+      this._handleEventSnapRemoved,
+    );
+
+    this.messagingSystem.subscribe(
+      'SnapController:snapUpdated',
+      this._handleEventSnapUpdated,
+    );
+  }
+
+  /**
+   * Retrieve all cronjob specifications for all runnable snaps.
+   *
+   * @returns Array of Cronjob specifications.
+   */
+  private async getAllJobs(): Promise<Cronjob[]> {
+    const snaps = await this.messagingSystem.call('SnapController:getAll');
+    const filteredSnaps = getRunnableSnaps(snaps);
+
+    const jobs = await Promise.all(
+      filteredSnaps.map((snap) => this.getSnapJobs(snap.id)),
+    );
+    return flatten(jobs).filter((job) => job !== undefined) as Cronjob[];
+  }
+
+  /**
+   * Retrieve all Cronjob specifications for a Snap.
+   *
+   * @param snapId - ID of a Snap.
+   * @returns Array of Cronjob specifications.
+   */
+  private async getSnapJobs(snapId: SnapId): Promise<Cronjob[] | undefined> {
+    const permissions = await this._messenger.call(
+      'PermissionController:getPermissions',
+      snapId,
+    );
+
+    const permission = permissions?.[SnapEndowments.Cronjob];
+    const definitions = getCronjobCaveatJobs(permission);
+
+    return definitions?.map((definition, idx) => {
+      return { ...definition, id: `${snapId}-${idx}`, snapId };
+    });
+  }
+
+  /**
+   * Register cron jobs for a given snap by getting specification from a permission caveats.
+   * Once registered, each job will be scheduled.
+   *
+   * @param snapId - ID of a snap.
+   */
+  async register(snapId: SnapId) {
+    const jobs = await this.getSnapJobs(snapId);
+    jobs?.forEach((job) => this.schedule(job));
+  }
+
+  /**
+   * Schedule job.
```

Comment:
Maybe expand on this a little bit, as this doesn't explain everything that the function is doing.

Reply:
Expanded in the last commit. It will explain a little bit more about the process. Let me know if you have any other suggestion, I'll add more if needed.